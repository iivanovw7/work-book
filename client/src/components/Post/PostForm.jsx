import { opacify } from 'polished';
import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import Button from '../../elements/UI/Button';
import { Input } from '../../elements/UI/Input';
import { Textarea } from '../../elements/UI/Textarea';
import { useInput } from '../../utils/useInputHook';
/* eslint no-underscore-dangle: 0 */
/* eslint react/require-default-props: 0 */

const StyledForm = styled.form`
	display: flex;
	flex-direction: column;
	align-items:flex-start;
	justify-content: center;
	
	code {
		color: red;
	}
`;

const ButtonsContainer = styled.div`
	width: 100%;
	margin-top: 1em;
	margin-bottom: 1em;
	display: flex;
	flex-direction: row;
	flex-wrap: no-wrap;
	justify-content: space-between;
	user-select: none;
	padding: 0.5rem;
	border-radius: 0.2em;
	background-color: ${opacify('0.01', 'rgba(25, 0, 0, 0.1)')};
`;


export const PostForm = (props) => {
	const {
		history, addPost, updatePost, data, theme, locale, text
	} = props;
	const postData = (data ? data.getPost : {});
	const { value: title, bind: bindTitle, reset: resetTitle } = useInput(postData.title || 'title');
	const { value: postText, bind: bindPostText, reset: resetPostText } = useInput(postData.text || 'text');
	const { value: subject, bind: bindSubject, reset: resetSubject } = useInput(postData.subject || 'subject');
	const { value: tags, bind: bindTags, reset: resetTags } = useInput(postData.tags || 'js,apollo,react');

	function resetForm() {
		resetTitle();
		resetPostText();
		resetSubject();
		resetTags();
	}

	const formVariables = {
		variables: {
			_id: postData ? postData._id : '',
			subject,
			text: postText,
			title,
			tags: postData ? tags : tags.split(',')
		}
	};

	const handleSubmit = (event) => {
		event.preventDefault();
		data ? updatePost(formVariables) : addPost(formVariables);
		resetForm();
		history.push('/posts');
	};

	return (
		<StyledForm onSubmit={handleSubmit}>
			<Input id="Title" type="text" label="Title" {...bindTitle} />
			<Input id="Subject" type="text" label="Subject" {...bindSubject} />
			<Textarea id="Text" type="text" label="Text" rows={40} {...bindPostText} />
			<Input id="Tags" type="text" label="Tags" {...bindTags} />
			<ButtonsContainer>
				<Button
					text={text.navigation.back[locale]}
					theme={theme}
					variant="primary"
					handleClick={() => {
						history.push('/');
					}}
				/>
				<Button
					text={text.navigation.save[locale]}
					variant="primary"
					theme={theme}
					type="submit"
					value="Submit"
				/>
			</ButtonsContainer>
		</StyledForm>
	);
};

PostForm.propTypes = {
	addPost: PropTypes.func,
	data: PropTypes.object,
	history: PropTypes.object.isRequired,
	theme: PropTypes.string.isRequired,
	updatePost: PropTypes.func,
	locale: PropTypes.string.isRequired,
	text: PropTypes.object.isRequired
};
