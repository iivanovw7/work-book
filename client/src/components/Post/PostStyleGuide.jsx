import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import ListItem from '../../elements/UI/LIstElement';
/* eslint no-underscore-dangle: 0 */
/* eslint react/require-default-props: 0 */

const List = styled.ul`
  text-decoration: none;
  list-style-type: none;
  margin-bottom: 0;
`;

const PostStyleGuide = (props) => {
  const { theme, variant, locale } = props;

  return (
    <List>
      <ListItem
        theme={theme}
        variant={variant}
        fontSize="0.6em"
        text={'<a target=\\"_blank\\" href="https://URL">Link</a>'}
        description="to insert link in post."
        locale={locale}
      />
      <ListItem
        theme={theme}
        variant={variant}
        fontSize="0.6em"
        text={'<code>text</code>'}
        description="to set background for text."
        locale={locale}
      />
      <ListItem
        theme={theme}
        variant={variant}
        fontSize="0.6em"
        text={'<pre class="prettyprint">code</pre>'}
        description="to create a code sample."
        locale={locale}
      />
      <ListItem
        theme={theme}
        variant={variant}
        fontSize="0.6em"
        text={'&lt; or &gt;'}
        description="to set < or > in code sample"
        locale={locale}
      />
    </List>
  );
};

PostStyleGuide.propTypes = {
  variant: PropTypes.string,
  theme: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired
};

export default PostStyleGuide;
