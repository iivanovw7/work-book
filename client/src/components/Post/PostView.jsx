import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import moment from 'moment';
import styled, { ThemeProvider } from 'styled-components';
import { opacify } from 'polished';
import ErrorMessage from '../../elements/errorMessage';
import Button from '../../elements/UI/Button';
import * as utils from '../../utils';
import { backgroundColorInverse, textColorInverse, linkColor } from '../../theme';
import DeletePost from '../../queryBlocks/DeletePost';
import TagsCloud from '../Tags/TagsCloud';
import TextLink from '../../elements/UI/TextLink';
/* eslint no-underscore-dangle: 0 */

const StyledPostContainer = styled.div`
	margin-bottom: 1em;
	
	code {
		background-color: ${backgroundColorInverse};
		opacity: 0.7;
		color: ${textColorInverse};
		padding: 0.1em;
	}
`;

const StyledText = styled.p`
	padding: 0.5rem;
	border-radius: 0.2em;
	background-color: ${opacify('0.01', 'rgba(25, 0, 0, 0.1)')};
	
	a {
	  color: ${linkColor};
	}
`;

const StyledTimeContainer = styled.p`
	font-size: 0.7em;
`;

const StyledButtonsContainer = styled.div`
	display: flex;
	flex-direction: row;
	flex-wrap: no-wrap;
	user-select: none;
	justify-content: space-between;
	padding: 0.5rem;
	border-radius: 0.2em;
	background-color: ${opacify('0.01', 'rgba(25, 0, 0, 0.1)')};
	
	div:first-of-type {
		display: flex;
		flex-direction: row;
	}
	
`;

const StyledTagsContainer = styled.div`
	display: flex;
`;

const PostView = (props) => {
  const {
    history, data, theme, locale, text
  } = props;
  const post = data.getPost;
  const date = post ? String(post.created) : 'Invalid date';
  const [userAccess, setUserAccess] = useState(false);
  const est = utils.calculateReadingTime(post.text.length, locale);

  useEffect(() => {
    const getAccessRights = async () => setUserAccess(await utils.checkUser());
    utils.runCodePrettify();
    getAccessRights()
      .catch(e => console.log(e));
  }, []);

  function handleClick(id) {
    history.push(`/posts/${id}`);
  }

  if (!data.getPost) {
    return (
      <ErrorMessage
        text="Back"
        theme={theme}
        message="Post not found!"
        handleClick={() => {
          history.push('/posts');
        }}
      />
    );
  }

  return (
    <ThemeProvider theme={{ mode: theme }} key={post._id}>
      <StyledPostContainer>
        <h2>
          <TextLink
            theme={theme}
            text={post.title}
            variant="secondary"
            link="/"
            handleClick={(e) => {
              e.preventDefault();
              handleClick(post._id);
            }}
          />
        </h2>
        <h3>
          {post.subject}
        </h3>
        <p className="dates">
          {moment(date, 'x')
            .format('DD MMM YYYY HH:MM A')}
        </p>
        <StyledTimeContainer className="dates">
          {est.minutes}
          {' '}
          {est.ending}
        </StyledTimeContainer>
        <StyledTagsContainer>
          <TagsCloud data={post.tags} tags={data.getPost.tags[0]} history={history} theme={theme} postTags {...props} />
        </StyledTagsContainer>
        <StyledText dangerouslySetInnerHTML={utils.sanitize(post.text)} />
        <StyledButtonsContainer>
          <Button
            variant="primary"
            text={text.navigation.back[locale]}
            theme={theme}
            maxWidth="7em"
            handleClick={() => {
              history.push('/');
            }}
          />
          <div>
            {userAccess && (
              <Button
                variant="primary"
                text={text.navigation.update[locale]}
                theme={theme}
                maxWidth="10em"
                handleClick={() => {
                  history.push(`/posts/update/${post._id}`);
                }}
              />
            )}
            {userAccess && (
              <DeletePost post={post} {...props} />
            )}
          </div>
        </StyledButtonsContainer>
      </StyledPostContainer>
    </ThemeProvider>
  );
};

export default PostView;

PostView.propTypes = {
  data: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  locale: PropTypes.string.isRequired,
  text: PropTypes.object.isRequired,
  theme: PropTypes.string.isRequired
};
