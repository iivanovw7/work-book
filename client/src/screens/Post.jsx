import PropTypes from 'prop-types';
import React from 'react';
import useStoreon from 'storeon/react';
import { withRouter } from 'react-router-dom';
import { Col, Grid, Row } from 'react-styled-flexboxgrid';
import styled, { ThemeProvider } from 'styled-components';
import SwitchedPostComponent from '../utils/postHelpers';
import MobileNavBar from '../components/MobileNavBar/MobileNavbar';
import TopBar from '../components/TopBar/TopBar';
import { backgroundColor, textColor } from '../theme';
import SideBar from '../components/SideBar/SideBar';
/* eslint react/require-default-props: 0 */
/* eslint no-underscore-dangle: 0 */

const StyledSection = styled.div`
	color: ${textColor};
  background-color: ${backgroundColor};
  min-width: 100%
  box-sizing: content-box;
`;

const StyledSideBar = styled.aside`
  max-width: inherit;
`;

const Post = (props) => {
  const { location } = props;
  const { theme, locale, localizedText } = useStoreon('theme', 'locale', 'localizedText');

  return (
    <ThemeProvider theme={{ mode: theme }}>
      <StyledSection className="container">
        <section>
          <TopBar locale={locale} text={localizedText} {...props} />
        </section>
        <Grid>
          <Row>
            <Col xs={12} sm={12} md={3} lg={3}>
              <StyledSideBar>
                <SideBar
                  theme={theme}
                  locale={locale}
                  text={localizedText}
                  onlyMobile
                  {...props}
                />
              </StyledSideBar>
            </Col>
            <Col xs={12} sm={12} md={9} lg={9}>
              <SwitchedPostComponent
                location={location}
                theme={theme}
                locale={locale}
                text={localizedText}
              />
            </Col>
          </Row>
        </Grid>
        <section className="mobile">
          <MobileNavBar theme={theme} locale={locale} text={localizedText} {...props} />
        </section>
      </StyledSection>
    </ThemeProvider>
  );
};

Post.propTypes = {
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  userAccess: PropTypes.bool.isRequired
};

export default withRouter(Post);
