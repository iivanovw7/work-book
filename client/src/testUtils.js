import { createMemoryHistory } from 'history';
import * as localized from './assets/locales.json';

// Mock browser history
export const history = createMemoryHistory('/dashboard');

// Grid config object for testing
export const mappedSizes = {
	xs: 12,
	sm: 12,
	md: 12,
	lg: 11
};

export const localizedText = {
	locales: localized.locales,
	topBar: localized.topBar,
	login: localized.login,
	navigation: localized.navigation,
	pageNotFound: localized.pageNotFound
};
