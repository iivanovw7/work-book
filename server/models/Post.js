import mongoose from 'mongoose';

const postSchema = mongoose.Schema({
	subject: String,
	text: String,
	title: String,
	tags: Array,
	author: { type: mongoose.Schema.ObjectId, ref: 'User' },
	created: { type: Date, default: Date.now() },
	published: { type: Boolean, default: false }
});

export default mongoose.model('Post', postSchema, 'posts');
