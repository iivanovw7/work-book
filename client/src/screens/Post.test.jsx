import React from 'react';
import chalk from 'chalk';
import { MemoryRouter } from 'react-router-dom';
import Post from './Post';
import * as testUtils from '../testUtils';

jest.mock('../config/apiURL', () => () => '/api');
jest.mock('storeon/react', () => () => ({
  locale: 'eng',
  theme: 'dark',
  search: ''
}));
jest.mock('../components/Navigation/MobileNavbar', () => () => (
  <div id="MobileNavbar" />
));
jest.mock('../components/Navigation/TopBar', () => () => (
  <div id="TopBar" />
));
jest.mock('../components/Navigation/SideBar', () => () => (
  <div id="SideBar" />
));
jest.mock('../utils/postHelpers', () => () => (
  <div id="SwitchedComponent" />
));

describe(`Testing [${chalk.yellow('POST')}] screen: `, () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  const props = {
    localizedText: testUtils.localizedText,
    location: {},
    user: {},
    userAccess: true
  };

  const Composition = () => (
    <MemoryRouter>
      <Post {...props} />
    </MemoryRouter>
  );

  it('Should render properly with props ', () => {
    const component = testUtils.suppressConsoleWarnings(
      <Composition />,
      'mount',
      console
    );

    expect(component.find('Post').length).toEqual(1);
    expect(component.find('#MobileNavbar').length).toEqual(1);
    expect(component.find('#TopBar').length).toEqual(1);
    expect(component.find('#SideBar').length).toEqual(1);
  });

  it('Should render and match snapshot ', () => {
    const component = testUtils.suppressConsoleWarnings(
      <Composition />,
      'mount',
      console
    );

    expect(component).toMatchSnapshot();
  });
});
