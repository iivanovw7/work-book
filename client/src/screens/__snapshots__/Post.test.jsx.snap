// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing [[33mPOST[39m] screen:  Should render and match snapshot  1`] = `ReactWrapper {}`;
